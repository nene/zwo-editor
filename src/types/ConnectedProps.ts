type AnyFn = (...args: any) => any;

type ReturnVoid<F extends AnyFn> = (...args: Parameters<F>) => void;

type FunctionMap = { [key: string]: AnyFn };

type VoidFunctionMap<T extends FunctionMap> = {
  [P in keyof T]: ReturnVoid<T[P]>;
};

// Derives Props type from mapStateToProps and mapDispatchToProps
export type ConnectedProps<
  MapStateToProps extends AnyFn,
  MapDispatchToProps extends FunctionMap = {}
> = ReturnType<MapStateToProps> & VoidFunctionMap<MapDispatchToProps>;
