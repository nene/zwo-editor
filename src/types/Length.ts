export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;

// Alternative approach
//
// This way we could avoid storing class instances to Redux,
// which eliminates problems with serialization & deserialization.
/*
export type Distance = { meters: number };
export type Duration = { seconds: number };
export type Length = Distance | Duration;

export const Distance = (meters: number) => ({ meters });
export const Duration = (seconds: number) => ({ seconds });

export const isDistance = (obj: Length): obj is Distance => 'meters' in obj;
export const isDuration = (obj: Length): obj is Duration => 'seconds' in obj;
*/
