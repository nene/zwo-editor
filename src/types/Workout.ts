import { SportType } from "./SportType";
import { Interval } from "./Interval";
import { LengthType } from "./LengthType";

export interface Workout {
  author: string;
  name: string;
  description: string;
  sportType: SportType;
  lengthType: LengthType;
  tags: string[];
  intervals: Array<Interval>;
}

export function createEmptyWorkout(
  sportType: SportType,
  lengthType: LengthType
): Workout {
  return {
    author: "",
    name: "",
    description: "",
    sportType: sportType,
    lengthType: lengthType,
    tags: [],
    intervals: [],
  };
}
