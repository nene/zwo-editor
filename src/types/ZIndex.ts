export enum ZIndex {
  fader = 1,
  selected = 5,
  instruction = 11,
  tooltip = 100,
  popup = 1000,
}
