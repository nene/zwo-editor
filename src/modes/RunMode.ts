import { SportType } from "../types/SportType";
import { PaceType } from "../types/PaceType";
import { RunningTimes } from "../types/RunningTimes";
import { runningDistances } from "../types/runningDistances";
import { LengthType } from "../types/LengthType";
import Mode from "./Mode";
import {
  Distance,
  Duration,
  isDistance,
  isDuration,
  Length,
} from "../types/Length";
import { Interval } from "../types/Interval";

export default class RunMode extends Mode {
  private runningTimes: RunningTimes;
  public readonly sportType: SportType = "run";
  public readonly lengthType: LengthType;

  constructor(runningTimes: RunningTimes, lengthType: LengthType) {
    super();
    this.runningTimes = runningTimes;
    this.lengthType = lengthType;
  }

  speed(intensity: number, pace: PaceType): number {
    return (runningDistances[pace] / this.runningTimes[pace]) * intensity; // in m/s
  }

  distance(length: Length, intensity: number, pace: PaceType): Distance {
    if (isDuration(length)) {
      return Distance(this.speed(intensity, pace) * length.seconds);
    } else {
      return length;
    }
  }

  duration(length: Length, intensity: number, pace: PaceType): Duration {
    if (isDuration(length)) {
      return length;
    } else {
      return Duration(length.meters / this.speed(intensity, pace));
    }
  }

  intervalDistance(interval: Interval): Distance {
    switch (interval.type) {
      case "free": {
        return isDistance(interval.length) ? interval.length : Distance(0);
      }
      case "steady": {
        return this.distance(
          interval.length,
          interval.intensity,
          interval.pace
        );
      }
      case "ramp": {
        return this.distance(
          interval.length,
          (interval.startIntensity + interval.endIntensity) / 2,
          interval.pace
        );
      }
      case "repetition": {
        const onDistance = this.distance(
          interval.onLength,
          interval.onIntensity,
          interval.pace
        );
        const offDistance = this.distance(
          interval.offLength,
          interval.offIntensity,
          interval.pace
        );
        return Distance(
          interval.repeat * (onDistance.meters + offDistance.meters)
        );
      }
    }
  }

  intervalDuration(interval: Interval): Duration {
    switch (interval.type) {
      case "free":
        return isDuration(interval.length) ? interval.length : Duration(0);
      case "steady":
        return this.duration(
          interval.length,
          interval.intensity,
          interval.pace
        );
      case "ramp":
        return this.duration(
          interval.length,
          (interval.startIntensity + interval.endIntensity) / 2,
          interval.pace
        );
      case "repetition": {
        const onDuration = this.duration(
          interval.onLength,
          interval.onIntensity,
          interval.pace
        );
        const offDuration = this.duration(
          interval.offLength,
          interval.offIntensity,
          interval.pace
        );
        return Duration(
          interval.repeat * (onDuration.seconds + offDuration.seconds)
        );
      }
    }
  }

  intervalLength(interval: Interval): Length {
    if (this.lengthType === "time") {
      return this.intervalDuration(interval);
    } else {
      return this.intervalDistance(interval);
    }
  }
}
