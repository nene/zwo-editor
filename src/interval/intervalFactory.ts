import { v4 as uuidv4 } from "uuid";
import { PaceType } from "../types/PaceType";
import {
  FreeInterval,
  Interval,
  RampInterval,
  RepetitionInterval,
  SteadyInterval,
} from "../types/Interval";
import { Distance, Duration, Length } from "../types/Length";
import { WorkoutMode } from "../modes/WorkoutMode";

const defaultDuration = Duration(300);
const defaultDistance = Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: "steady",
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: "ramp",
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: "free",
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: undefined,
      instructions: [],
      ...interval,
    };
  },

  repetition(
    interval: Partial<RepetitionInterval>,
    mode: WorkoutMode
  ): RepetitionInterval {
    return {
      type: "repetition",
      id: uuidv4(),
      repeat: 3,
      onLength: mode.lengthType === "time" ? Duration(30) : Distance(200),
      offLength: mode.lengthType === "time" ? Duration(120) : Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      onCadence: undefined,
      offCadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
