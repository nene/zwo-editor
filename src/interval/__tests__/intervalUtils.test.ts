import {
  moveInterval,
  updateIntervalLength,
  updateIntervalIntensity,
} from "../intervalUtils";
import { Distance, Duration } from "../../types/Length";
import intervalFactory from "../intervalFactory";
import BikeMode from "../../modes/BikeMode";
import { WorkoutMode } from "../../modes/WorkoutMode";
import RunMode from "../../modes/RunMode";

const defaultBikeMode = () => new BikeMode(200, 75);
const distanceRunMode = () => new RunMode([], "distance");

function deepFreeze<T>(object: T): T {
  for (const name of Object.getOwnPropertyNames(object)) {
    const value = (object as any)[name];
    if (value && typeof value === "object") {
      deepFreeze(value);
    }
  }
  return Object.freeze(object);
}

// Use freeze to ensure we aren't accidentally mutating our interval list
const defaultIntervals = (mode: WorkoutMode) =>
  deepFreeze([
    intervalFactory.steady(
      {
        id: "#1",
        length: Duration(60),
        intensity: 0.5,
      },
      mode
    ),
    intervalFactory.steady(
      {
        id: "#2",
        length: Duration(120),
        intensity: 1.0,
      },
      mode
    ),
    intervalFactory.steady(
      {
        id: "#3",
        length: Duration(30),
        intensity: 0.75,
      },
      mode
    ),
  ]);

const defaultRunIntervals = (mode: WorkoutMode) =>
  deepFreeze([
    intervalFactory.steady(
      {
        id: "#1",
        length: Distance(1000),
        intensity: 0.5,
      },
      mode
    ),
    intervalFactory.steady(
      {
        id: "#2",
        length: Distance(1200),
        intensity: 1.0,
      },
      mode
    ),
    intervalFactory.steady(
      {
        id: "#3",
        length: Distance(3000),
        intensity: 0.75,
      },
      mode
    ),
  ]);

describe("intervalUtils", () => {
  describe("updateIntervalLength", () => {
    it("increases or decreses duration of specified interval", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(
        updateIntervalLength("#2", Duration(10), intervals, mode)
      ).toMatchSnapshot();
      expect(
        updateIntervalLength("#3", Duration(-10), intervals, mode)
      ).toMatchSnapshot();
    });

    it("increases or decreses distance of specified interval", () => {
      const mode = distanceRunMode();
      const intervals = defaultRunIntervals(mode);
      expect(
        updateIntervalLength("#2", Distance(100), intervals, mode)
      ).toMatchSnapshot();
      expect(
        updateIntervalLength("#3", Distance(-100), intervals, mode)
      ).toMatchSnapshot();
    });

    it("does not allow decreasing duration to zero or below it", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(
        updateIntervalLength("#1", Duration(-65), intervals, mode)
      ).toEqual(intervals);
      expect(
        updateIntervalLength("#1", Duration(-60), intervals, mode)
      ).toEqual(intervals);
    });

    it("does not allow decreasing distance to zero or below it", () => {
      const mode = distanceRunMode();
      const intervals = defaultRunIntervals(mode);
      expect(
        updateIntervalLength("#1", Distance(-1200), intervals, mode)
      ).toEqual(intervals);
      expect(
        updateIntervalLength("#1", Distance(-1000), intervals, mode)
      ).toEqual(intervals);
    });

    it("does nothing when ID not found", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(
        updateIntervalLength("#blah", Duration(10), intervals, mode)
      ).toEqual(intervals);
    });
  });

  describe("updateIntervalIntensity", () => {
    it("increases or decreses duration of specified interval", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalIntensity("#2", 0.05, intervals)).toMatchSnapshot();
      expect(updateIntervalIntensity("#3", -0.05, intervals)).toMatchSnapshot();
    });

    it("does not allow decreasing intensity to zero or below it", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalIntensity("#1", -0.55, intervals)).toEqual(
        intervals
      );
      expect(updateIntervalIntensity("#1", -0.5, intervals)).toEqual(intervals);
    });

    it("does nothing when ID not found", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(updateIntervalIntensity("#blah", 0.1, intervals)).toEqual(
        intervals
      );
    });
  });

  describe("moveInterval", () => {
    it("moves interval index by +1 or -1", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(moveInterval("#2", -1, intervals)).toMatchSnapshot();
      expect(moveInterval("#2", +1, intervals)).toMatchSnapshot();
    });

    it("does not allow moving past start or end", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(moveInterval("#1", -1, intervals)).toEqual(intervals);
      expect(moveInterval("#3", +1, intervals)).toEqual(intervals);
    });

    it("does nothing when ID not found", () => {
      const mode = defaultBikeMode();
      const intervals = defaultIntervals(mode);
      expect(moveInterval("#blah", 1, intervals)).toEqual(intervals);
    });
  });
});
