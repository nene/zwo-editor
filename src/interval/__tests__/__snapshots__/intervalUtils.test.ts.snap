// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intervalUtils moveInterval moves interval index by +1 or -1 1`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#2",
    "instructions": Array [],
    "intensity": 1,
    "length": Object {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#3",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils moveInterval moves interval index by +1 or -1 2`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#3",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#2",
    "instructions": Array [],
    "intensity": 1,
    "length": Object {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalIntensity increases or decreses duration of specified interval 1`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#2",
    "instructions": Array [],
    "intensity": 1.05,
    "length": Object {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#3",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalIntensity increases or decreses duration of specified interval 2`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#2",
    "instructions": Array [],
    "intensity": 1,
    "length": Object {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#3",
    "instructions": Array [],
    "intensity": 0.7,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalLength increases or decreses distance of specified interval 1`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "meters": 1000,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#2",
    "instructions": Array [],
    "intensity": 1,
    "length": Object {
      "meters": 1300,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#3",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "meters": 3000,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalLength increases or decreses distance of specified interval 2`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "meters": 1000,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#2",
    "instructions": Array [],
    "intensity": 1,
    "length": Object {
      "meters": 1200,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#3",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "meters": 2900,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalLength increases or decreses duration of specified interval 1`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#2",
    "instructions": Array [],
    "intensity": 1,
    "length": Object {
      "seconds": 130,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#3",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "seconds": 30,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;

exports[`intervalUtils updateIntervalLength increases or decreses duration of specified interval 2`] = `
Array [
  Object {
    "cadence": undefined,
    "id": "#1",
    "instructions": Array [],
    "intensity": 0.5,
    "length": Object {
      "seconds": 60,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#2",
    "instructions": Array [],
    "intensity": 1,
    "length": Object {
      "seconds": 120,
    },
    "pace": 0,
    "type": "steady",
  },
  Object {
    "cadence": undefined,
    "id": "#3",
    "instructions": Array [],
    "intensity": 0.75,
    "length": Object {
      "seconds": 20,
    },
    "pace": 0,
    "type": "steady",
  },
]
`;
